### PR TITLE
Deployment adds revision info to its new rc

### PR DESCRIPTION
**Pull request by janetkuo**
_Updated: we decided to rename version to "revision"._
_Updated: we decided to rename the revision annotation to "deployment.kubernetes.io/revision"_

---

Addresses #17168; depends on #20139 

Use annotation `deployment.kubernetes.io/revision` as a revision indicator when deployment creates its new rcs. It will be used for rolling back a deployment and for viewing deployment rollout history.  

See [proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deploy.md#deployment-version):

> `rollout.kubectl.kubernetes.io/version` [renamed to `deployment.kubernetes.io/revision`] records a version number to distinguish the change sequence of a deployment's replication controllers. A deployment obtains the largest version number from its replication controllers and increments the number by 1 upon update or creation of the deployment, and update the version annotation of its new replication controller. 

Note that this annotation is stored in deployment and its RC. 

cc @bgrant0607 @nikhiljindal @ironcladlou @kargakis @kubernetes/sig-config

### CODE REVIEW

**Comment by kargakis** [comment]
How does this address https://github.com/kubernetes/kubernetes/issues/17168? I thought we were going to remove all the client code and use rolling deployments to facilitate updates.

**Comment by janetkuo** [comment]
@kargakis this is the first step for supporting rollback and view rollout history. We're not going to modify `kubectl rolling-update` directly but create a new command (`kubectl rollout`) which provides similar experience with Deployment.

**Comment by janetkuo** [comment]
Rebased. PTAL

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -557,9 +597,23 @@ func (dc *DeploymentController) getNewRC(deployment extensions.Deployment) (*api
...
+	}
+
 	return createdRC, nil
 }
 
+func (dc *DeploymentController) updateDeploymentVersion(deployment extensions.Deployment, version string) error {
+	if deployment.Annotations == nil {
+		deployment.Annotations = make(map[string]string)
+	}
+	deployment.Annotations[deploymentutil.VersionAnnotation] = version
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Shouldn't we do generation numbers instead of maintaining a version annotation? @bgrant0607

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Deployment version is from observing current RCs but not from this deployment version annotation. We look at the deployment's RCs, get the max version number among them, determine what version number we should give to the new RC, and update deployment version number as a reference. 

Even if we implement and use deployment generation numbers, we cannot count on it; otherwise, when the deployment is killed and then re-created (generation number becomes 1) we lost the actual version number.

**Comment by bgrant0607** [comment]
I know I agreed to "version" in the [proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deploy.md), but I would like to change to "revision" if nobody has any objection, here and everywhere, so that it's not confused with API versions, resourceVersion, Kubernetes release version, application version, etc.

Regardless which we agree on, we should make RevisionHistoryLimit match.

**Comment by bgrant0607** [comment]
Renamed #17168 to be more accurate given the current plan.

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: We also need generation and `observedGeneration`, similar to RC.

I haven't looked at the RC implementation, but `observedGeneration` should be updated once the changed made in the corresponding `generation` have made to the underlying resources (RCs in this case).

But versions/revisions have nothing to do with that. This is about creating understandable history.

Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Disadvantages of using generation for this purpose:
* generation needs to be incremented when changes are made that won't cause rollouts, such as scaling (changing replicas)
* generation would reset to 0 after deleting and re-creating the Deployment

Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: > generation needs to be incremented when changes are made that won't cause rollouts, such as scaling (changing replicas)

@bprashanth is this how generation is working for rcs? I think it is increased only on podtemplate changes.

Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: > generation would reset to 0 after deleting and re-creating the Deployment

The deployment controller could observe any adoptions and set the correct generation number back in the deployment.

Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: The reason we added generation/observedGeneration was to determine when the controller manager had observed replicas changes.

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/controller/strategy.go#L73
	// Any changes to the spec increment the generation number, any changes to the
	// status should reflect the generation number of the corresponding object. We push
	// the burden of managing the status onto the clients because we can't (in general)
	// know here what version of spec the writer of the status has seen. It may seem like
	// we can at first -- since obj contains spec -- but in the future we will probably make
	// status its own object, and even if we don't, writes may be the result of a
	// read-update-write loop, so the contents of spec may not actually be the spec that
	// the controller has *seen*.
	//
	// TODO: Any changes to a part of the object that represents desired state (labels,
	// annotations etc) should also increment the generation.
	if !reflect.DeepEqual(oldController.Spec, newController.Spec) {
		newController.Generation = oldController.Generation + 1
	}

Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Ah, I missed that it's the whole spec and not just the template.

**Comment by janetkuo** [comment]
Agree on changing version to "revision".

**Comment by janetkuo** [comment]
Renamed all reference to "version" to "revision". PTAL.

File: pkg/util/deployment/deployment.go [hunk 2]
```diff
@@ -28,6 +28,11 @@ import (
 	podutil "k8s.io/kubernetes/pkg/util/pod"
 )
 
+const (
+	// The revision annotation of a deployment's replication controllers which records its rollout sequence
+	RevisionAnnotation = "rollout.kubectl.kubernetes.io/revision"
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: These revisions aren't generated by kubectl. They are generated by deployment, so deployment.kubernetes.io/revision would be more appropriate.

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -507,6 +496,51 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
...
+	}
+	return newRC, oldRCs, nil
+}
+
+func revision(rc *api.ReplicationController) (int, error) {
+	v, ok := rc.Annotations[deploymentutil.RevisionAnnotation]
+	if !ok {
+		return 0, nil
+	}
+	return strconv.Atoi(v)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Atoi could also return an error. We should handle that.

File: pkg/controller/deployment/deployment_controller.go [hunk 4]
```diff
@@ -507,6 +496,51 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
...
+}
+
+func maxRevision(allRCs []*api.ReplicationController) (int, error) {
+	max := 0
+	for _, rc := range allRCs {
+		if v, err := revision(rc); err != nil {
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: I don't think we should bail here in the case that one RC isn't properly annotated. We should degrade gracefully -- just skip it. It won't appear in the revision history in kubectl and perhaps it wouldn't be possible to automatically roll back to that revision, but all the other behavior should work correctly.

File: pkg/controller/deployment/deployment_controller.go [hunk 5]
```diff
@@ -507,6 +496,51 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
...
+		return nil, nil, err
+	}
+
+	newRC, err := dc.getNewRC(deployment, maxOldV)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if newRC.Annotations != nil && newRC.Annotations[deploymentutil.RevisionAnnotation] == "" &&
+		deployment.Annotations == nil || deployment.Annotations[deploymentutil.RevisionAnnotation] != newRC.Annotations[deploymentutil.RevisionAnnotation] {
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: `||` has the lowest operator precedence, so the last relational expression will be evaluated in the case that any of the preceding ones fail, which isn't right. The expressions on this line need to be parenthesized.

File: pkg/util/deployment/deployment.go (hunk 2 above)
Comment (janetkuo) [inline pkg/util/deployment/deployment.go]: Changed to `deployment.kubernetes.io/revision`.

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: As discussed the error is returned too.

File: pkg/controller/deployment/deployment_controller.go (hunk 4 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Done. Log this error and continue the calculation.

File: pkg/controller/deployment/deployment_controller.go (hunk 5 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Fixed

**Comment by janetkuo** [comment]
All comments addressed. I made the new RC revision to be updated periodically. I also include old RCs with no pods when calculating max revision number (old RCs with no pods were omitted when reconciling for deployment rolling-update). 

I also make deployment's rcStore list namespaced (filed a separate PR #20139). 

PTAL

File: pkg/controller/deployment/deployment_controller.go [hunk 6]
```diff
@@ -507,26 +496,85 @@ func (dc *DeploymentController) syncDeploymentStatus(allRCs []*api.ReplicationCo
...
+	if err != nil {
+		return nil, err
+	} else if existingNewRC != nil {
+		if existingNewRC.Annotations == nil {
+			existingNewRC.Annotations = make(map[string]string)
+		}
+		if existingNewRC.Annotations[deploymentutil.RevisionAnnotation] != newRevision {
+			existingNewRC.Annotations[deploymentutil.RevisionAnnotation] = newRevision
+			glog.V(4).Infof("update existingNewRC's revision to %s - %+v\n", newRevision, existingNewRC)
+			return dc.client.ReplicationControllers(deployment.ObjectMeta.Namespace).Update(existingNewRC)
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Won't these updates conflict with scaling of the controller?

**Comment by krmayankk** [comment]
Is it possible to start testing these features by downloading the latest master repo or are these changes in a specific branch and would it be stable enough to do basic deployment or should I expect crashes 😀

File: pkg/controller/deployment/deployment_controller.go (hunk 6 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: The RC's revision annotation is updated before it get scaled. If there's update conflict it'll retry later before scaling it.

**Comment by janetkuo** [comment]
@krmayankk these features aren't (but will be) merged to master branch.

**Comment by bgrant0607** [comment]
@krmayankk I'd wait a week or so for all the outstanding changes to land.

File: test/e2e/deployment.go [hunk 7]
```diff
@@ -34,13 +34,13 @@ var _ = Describe("Deployment", func() {
 	It("deployment should create new pods", func() {
 		testNewDeployment(f)
 	})
-	It("RollingUpdateDeployment should delete old pods and create new ones [Flaky]", func() {
+	It("deployment should delete old pods and create new ones with RollingUpdateDeployment strategy [Flaky]", func() {
```
Comment (bgrant0607) [inline test/e2e/deployment.go]: We have a lot of PRs in flight. Please avoid unrelated changes like this, since they'll likely cause other PRs to need rebases.

**Comment by bgrant0607** [comment]
@janetkuo Looks pretty good now, except for the unrelated change in e2e.

Does #20139 need to merge first?

**Comment by janetkuo** [comment]
@bgrant0607 yes, #20139 needs to merge first.

File: test/e2e/deployment.go (hunk 7 above)
Comment (janetkuo) [inline test/e2e/deployment.go]: Will remove this change.

File: pkg/controller/deployment/deployment_controller.go (hunk 6 above)
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: I think it's fine since the update call will return the rc with the latest resource version

File: pkg/util/deployment/deployment.go [hunk 8]
```diff
@@ -28,8 +28,13 @@ import (
 	podutil "k8s.io/kubernetes/pkg/util/pod"
 )
 
+const (
+	// The revision annotation of a deployment's replication controllers which records its rollout sequence
+	RevisionAnnotation = "deployment.kubernetes.io/revision"
```
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: asked about this in https://github.com/kubernetes/kubernetes/pull/19686#discussion_r50973865 but I think this is the correct place

**Comment by kargakis** [comment]
one question, lgtm otherwise

File: pkg/util/deployment/deployment.go (hunk 8 above)
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: You mean is not the correct place?

**Comment by bgrant0607** [comment]
Please test that an adopted RC without a revision annotation does what we expect.

File: pkg/util/deployment/deployment.go (hunk 8 above)
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: > You mean is not the correct place?

Sorry for being ambiguous. I meant that this is the correct place to ask because this is the original PR making the change.

File: test/e2e/deployment.go (hunk 7 above)
Comment (janetkuo) [inline test/e2e/deployment.go]: Done

**Comment by janetkuo** [comment]
> Please test that an adopted RC without a revision annotation does what we expect.

In test/e2e/deployment.go, `testRollingUpdateDeployment` tests that: an old RC without revision annotation being scaled down by a deployment, a new RC is created as revision 1, and the deployment becomes revision 1. 

@bgrant0607 the unrelated change in e2e test is removed. PTAL

File: pkg/util/deployment/deployment.go (hunk 8 above)
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: I agree that all the deployment-related annotations belong in the same place, but it's ok with me if this gets moved in a subsequent PR, since we have so many interdependent PRs right now.

**Comment by bgrant0607** [comment]
LGTM but we need  #20139 to merge first, so not applying the label to this one yet.

File: pkg/util/deployment/deployment.go (hunk 8 above)
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: fine with me

**Comment by janetkuo** [comment]
@bgrant0607 I just updated the code to fix a bug that ListOptions isn't applied when Listing RCs. PTAL.

**Comment by bgrant0607** [comment]
LGTM

**Comment by mikedanese** [comment]
@jvalencia this might be of interest to you. It adds revision history to deployment as descirbed [here](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deploy.md#deployment-version).

**Comment by janetkuo** [comment]
#20139 is in. Applying LGTM
